[Computed Style Gen] Support serializers for computed values that go directly from RenderStyle to String

weinig opened the pull request:
#### 7240a7b5486e3577f32f8fc0f8e5ce4c0d78248f
<pre>
[Computed Style Gen] Support serializers for computed values that go directly from RenderStyle to String
<a href="https://bugs.webkit.org/show_bug.cgi?id=293301">https://bugs.webkit.org/show_bug.cgi?id=293301</a>

Reviewed by Darin Adler.

Adds support for `CSSComputedStyleDeclaration::getPropertyValue` that
implements serialization of properties without a temporary CSSValue.

Instead, we now have parallel code for both conversion from a RenderStyle
value to CSSValue and serialization of a RenderStyle value to string.

The majority of property serializers are generated (the same set of
properties as have CSSValue conversion generated), with the remaining
requiring custom implementations. This introduces a bit of duplicated
logic, but we will be able to collapse it (along with the logic in
ShorthandSerializer which it also mirrors), over time.

A few of the more involved properties have their serializers temporarily
implemented by using the a CSSValue conversion with a FIXME. These will be
updated in a subsequent change.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/style/StyleOrderedNamedLinesCollector.h: Added.
* Source/WebCore/style/StyleExtractorConverter.h:
    - Moved out of StyleExtractorConverter.h, as it didn&apos;t belong there.

* Source/WebCore/style/StyleExtractorSerializer.h: Added.
    - Implements analogues of Style::ExtractorConverter functions but
      for direct serialization.

* Source/WebCore/style/StyleBuilderConverter.h:
    - Remove dead forward declaration.

* Source/WebCore/style/StyleExtractorGenerated.h:
    - Add forward declaration for new generated serializer dispatch function.

* Source/WebCore/style/StyleExtractor.cpp:
    - Update to call new generated serializer dispatch function. As a debugging
      aid, a macro can be enabled that validates the direct serialization against
      serialization that goes through CSSValue.

* Source/WebCore/style/StyleExtractorCustom.h:
    - Add support for custom computed style direct serializers, mirroring
      conversion functions.

* Source/WebCore/css/CSSProperties.json:
    - Renamed a few shorthand patterns to better reflect what they were:
        2Value -&gt; CoalescingPair
        4Value -&gt; CoalescingQuad
        Standard -&gt; StandardSpaceSeparated
    - Added new shorthand pattern: StandardSlashSeparated
    - Resolved a few FIXMEs by adopting shorthand patterns

* Source/WebCore/css/scripts/process-css-properties.py:
    - Add support for a new serializers and serializer dispatch function
      in StyleExtractorGenerated.cpp

* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleExtractorGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StylePropertyShorthandFunctions.cpp:
    - Update shorthand pattern tests and fix typo. Regenerate results.

Canonical link: <a href="https://commits.webkit.org/295391@main">https://commits.webkit.org/295391@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d182a29546ef2a64ce99022ab4fbe2b2cc2f7748

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104930 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24644 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15065 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110146 "Built successfully") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106971 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25046 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33188 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79672 "Build is in progress. Recent messages:Running configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107936 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19482 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94701 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59979 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12779 "Build is in progress. Recent messages:Running configuration; 19 flakes 2 failures; Uploaded test results; Running re-run-layout-tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54987 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Running compile-webkit") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88943 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12825 "Build is in progress. Recent messages:Running configuration; Running run-layout-tests-in-stress-mode; Running layout-tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112632 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32095 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23605 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88748 "Build is in progress. Recent messages:Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; Running layout-tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32460 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90927 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88379 "Build is in progress. Recent messages:run-api-tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22529 "Built successfully and passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33274 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Running apply-patch; Running checkout-pull-request; Skipped layout-tests; Passed layout tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11054 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running run-layout-tests-in-stress-mode; Running layout-tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27410 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Running checkout-pull-request; Running compile-webkit") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32020 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37388 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31812 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35153 "Built successfully") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33371 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/style/StyleExtractorConverter.h:
```diff
@@ -592,7 +593,7 @@ inline RefPtr<CSSValue> ExtractorConverter::convertTransformOperation(const Rend
         return CSSFunctionValue::create(CSSValueRotateX, CSSPrimitiveValue::create(uncheckedDowncast<RotateTransformOperation>(operation).angle(), CSSUnitType::CSS_DEG));
     case TransformOperation::Type::RotateY:
-        return CSSFunctionValue::create(CSSValueRotateX, CSSPrimitiveValue::create(uncheckedDowncast<RotateTransformOperation>(operation).angle(), CSSUnitType::CSS_DEG));
+        return CSSFunctionValue::create(CSSValueRotateY, CSSPrimitiveValue::create(uncheckedDowncast<RotateTransformOperation>(operation).angle(), CSSUnitType::CSS_DEG));
```
This deserves a test case. This was found by the "check direct serialization against value serialization mode".

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserCustom.h:
```diff
@@ -123,10 +123,10 @@ class PropertyParserCustom {
     static bool consumeSingleShorthand(CSSParserTokenRange&, PropertyParserState&, const StylePropertyShorthand&, PropertyParserResult&);
-    static bool consume2ValueShorthand(CSSParserTokenRange&, PropertyParserState&, const StylePropertyShorthand&, PropertyParserResult&);
-    static bool consume4ValueShorthand(CSSParserTokenRange&, PropertyParserState&, const StylePropertyShorthand&, PropertyParserResult&);
+    static bool consumeCoalescingPairShorthand(CSSParserTokenRange&, PropertyParserState&, const StylePropertyShorthand&, PropertyParserResult&);
+    static bool consumeCoalescingQuadShorthand(CSSParserTokenRange&, PropertyParserState&, const StylePropertyShorthand&, PropertyParserResult&);
```
Are these terms “coalescing pair” and “coalescing quad” from the CSS specification or are they terms we coined for WebKit?

weinig replied:
I just coined them.

darinadler commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -203,7 +305,23 @@ template<typename MappingFunctor> Ref<CSSValue> extractAnimationOrTransitionValu
+            mapper(state, builder, context, includeComma, animation.ptr(), animationList);
+
+            if (builder.length() != lengthBefore)
+                includeComma = true;
```
I’d find this more readable without the blank lines.

darinadler commented on Source/WebCore/style/StyleExtractorCustom.h:
```diff
@@ -571,6 +821,12 @@ template<GridTrackSizingDirection direction> static Ref<CSSValue> extractGridTem
 
+template<GridTrackSizingDirection direction> void extractGridTemplateSerialization(ExtractorState& state, StringBuilder& builder, const CSS::SerializationContext& context)
+{
+    // FIXME: Optimize out CSSValue conversion.
```
```suggestion
   // FIXME: Do this more efficiently without creating and destroying a CSSValue object.
```

darinadler commented on Source/WebCore/style/StyleOrderedNamedLinesCollector.h:
```diff
@@ -0,0 +1,175 @@
+
+    void collectLineNamesForIndex(Vector<String>&, unsigned index) const override;
+
+    int namedGridLineCount() const override { return m_totalLines; }
```
Mix of `int` and `unsigned` here is not great.

darinadler commented on Source/WebCore/style/StyleOrderedNamedLinesCollector.h:
```diff
@@ -0,0 +1,175 @@
+    void collectLineNamesForIndex(Vector<String>&, unsigned index) const override;
+
+    int namedGridLineCount() const override { return m_totalLines; }
+private:
```
```suggestion

private:
```

darinadler commented on Source/WebCore/style/StyleOrderedNamedLinesCollector.h:
```diff
@@ -0,0 +1,175 @@
+    int namedGridLineCount() const override { return m_totalLines; }
+private:
+    unsigned m_insertionPoint;
+    unsigned m_autoRepeatTotalLineSets;
```
```suggestion
    unsigned m_autoRepeatTotalLineSets { 0 };
```
Then we could simplify the constructor.

weinig committed:
[Computed Style Gen] Support serializers for computed values that go directly from RenderStyle to String
https://bugs.webkit.org/show_bug.cgi?id=293301

Reviewed by Darin Adler.

Adds support for `CSSComputedStyleDeclaration::getPropertyValue` that
implements serialization of properties without a temporary CSSValue.

Instead, we now have parallel code for both conversion from a RenderStyle
value to CSSValue and serialization of a RenderStyle value to string.

The majority of property serializers are generated (the same set of
properties as have CSSValue conversion generated), with the remaining
requiring custom implementations. This introduces a bit of duplicated
logic, but we will be able to collapse it (along with the logic in
ShorthandSerializer which it also mirrors), over time.

A few of the more involved properties have their serializers temporarily
implemented by using the a CSSValue conversion with a FIXME. These will be
updated in a subsequent change.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/style/StyleOrderedNamedLinesCollector.h: Added.
* Source/WebCore/style/StyleExtractorConverter.h:
    - Moved out of StyleExtractorConverter.h, as it didn't belong there.

* Source/WebCore/style/StyleExtractorSerializer.h: Added.
    - Implements analogues of Style::ExtractorConverter functions but
      for direct serialization.

* Source/WebCore/style/StyleBuilderConverter.h:
    - Remove dead forward declaration.

* Source/WebCore/style/StyleExtractorGenerated.h:
    - Add forward declaration for new generated serializer dispatch function.

* Source/WebCore/style/StyleExtractor.cpp:
    - Update to call new generated serializer dispatch function. As a debugging
      aid, a macro can be enabled that validates the direct serialization against
      serialization that goes through CSSValue.

* Source/WebCore/style/StyleExtractorCustom.h:
    - Add support for custom computed style direct serializers, mirroring
      conversion functions.

* Source/WebCore/css/CSSProperties.json:
    - Renamed a few shorthand patterns to better reflect what they were:
        2Value -> CoalescingPair
        4Value -> CoalescingQuad
        Standard -> StandardSpaceSeparated
    - Added new shorthand pattern: StandardSlashSeparated
    - Resolved a few FIXMEs by adopting shorthand patterns

* Source/WebCore/css/scripts/process-css-properties.py:
    - Add support for a new serializers and serializer dispatch function
      in StyleExtractorGenerated.cpp

* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleExtractorGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StylePropertyShorthandFunctions.cpp:
    - Update shorthand pattern tests and fix typo. Regenerate results.

Canonical link: https://commits.webkit.org/295391@main
